### PR TITLE
Reconfiguration to snake

### PR DESCRIPTION
**Pull request by Trphajzlik**

### CODE REVIEW

**Comment by Trphajzlik** [comment]
Trochu jsem upravil svoji část kódu, rád přivítám nějaké návrhy na přerozdělení funkcí a struktur z snake_reconfig/*.h + zelegantnění minMaxHeap::popMin (/Max).

File: RoFILib/configuration/Generators.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
+#include <Configuration.h>
+#include <optional>
+#include <unordered_set>
+#include <vector>
+
+#ifndef ROBOTS_GENERATORS_H
+#define ROBOTS_GENERATORS_H
+
+    std::optional<Configuration> executeIfValid(const Configuration& config, const Action &action);
```
Comment (yaqwsx) [inline RoFILib/configuration/Generators.h:9]: Why the indent in the whole file?

File: RoFILib/generate/main.cpp [hunk 2]
```diff
@@ -102,17 +103,17 @@ ConfigPair generateTest(unsigned modules, unsigned path, unsigned step, unsigned
     {
         std::vector<Action::Rotate> rotate;
         std::vector<Action::Reconnect> reconnect;
-        goal.generateReconnect(reconnect);
-        goal.generateRotations(rotate, step);
+        generateReconnect(goal, reconnect);
+        generateRotations(goal, rotate, step);
         std::uniform_int_distribution<unsigned long> flip(0, 5);
         //if ((flip(e) > 1))
```
Comment (yaqwsx) [inline RoFILib/generate/main.cpp:109]: What is the reason for commented-out if?

File: RoFILib/reconfig/Algorithms.h [hunk 3]
```diff
@@ -70,26 +62,22 @@ class ConfigPool
...
-    {
+    Configuration* insert(const Configuration& config) {
         auto tmp = std::make_unique<Configuration>(config);
         auto [ptr, _] = pool.insert(std::move(tmp));
         return ptr->get();
     }
 
-    bool find(const Configuration& config) const
-    {
+    bool find(const Configuration& config) const {
```
Comment (yaqwsx) [inline RoFILib/reconfig/Algorithms.h:75]: Find is rather confusing name in this context regarding the return value - I would expect to find returns the object or index. Have you considered, e.g, `has`? `pool.has(configuration)` seem pretty idiomatic to me.

File: RoFILib/reconfig/astar.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    // Already computed shortest distance from init to configuration.
+    ConfigValue initDist;
+
+    // Shortest distance from init through this configuration to goal.
+    ConfigValue goalDist;
+
+    if (init == goal)
+    {
+        return {init};
+    }
```
Comment (yaqwsx) [inline RoFILib/reconfig/astar.cpp:19]: Remove the parenthesis to be consistent.

File: RoFILib/reconfig/astar.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+    std::priority_queue<EvalPair, std::vector<EvalPair>, EvalCompare> queue;
+
+    const Configuration* pointer = pool.insert(init);
```
Comment (yaqwsx) [inline RoFILib/reconfig/astar.cpp:23]: pointer is a poor name

File: RoFILib/reconfig/astar.cpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
+#include "Algorithms.h"
+
```
Comment (yaqwsx) [inline RoFILib/reconfig/astar.cpp:2]: Not sure if you want to refactor this file or let it be - I am ok with both.

File: RoFILib/reconfig/astar.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                auto path = createPath(pred, pointerNext);
+                if (stat != nullptr)
+                {
+                    stat->pathLength = path.size();
+                    stat->queueSize = maxQSize;
+                    stat->seenCfgs = pool.size();
+                }
+
+                return path;
+            }
```
Comment (yaqwsx) [inline RoFILib/reconfig/astar.cpp:78]: The body is quite long and it would be nice to factor some of the code into a function

File: RoFILib/reconfig/rrt.cpp [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+        --count;
+    }
+    if (count == 0)
+    {
+        return -1;
+    }
+    return index;
+}
+
+std::optional<Edge> generateEdge(ID id1, ID id2, std::unordered_map<ID, std::array<bool, 6>>& occupied)
```
Comment (yaqwsx) [inline RoFILib/reconfig/rrt.cpp:36]: If I am not mistaken, `generateEdge` and `generateRotation` are RRT local functions -> add static to them.

File: RoFILib/reconfig/rrt.cpp [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+        auto diff = curr.diff(goal);
+        return diff.rotations().size();
+    }
+
+    inline double diff(const Configuration& curr, const Configuration& goal) {
+        return reconnections(curr, goal) * 1000 + rotations(curr, goal);
+    }
+}
+
+inline int findFreeIndex(int index, const std::array<bool, 6>& A)
```
Comment (yaqwsx) [inline RoFILib/reconfig/rrt.cpp:21]: Not inline, but static (we are in a cpp file) - the same applies to other functions in this file. But a much better solution is to place them in an anonymous namespace.

File: RoFILib/reconfig/rrt.cpp (hunk 8 above)
Comment (yaqwsx) [reply RoFILib/reconfig/rrt.cpp:36]: Taking back. They need to be visible for tests.

File: RoFILib/snake_reconfig/MinMaxHeap.h [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+        } else {
+            std::cout << "no odd";
+        }
+        std::cout << "\nOccupied " << _occupied;
+        std::cout << "\n%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%" << std::endl;
+    }
+
+private:
+
+    inline int _l(int i) const {
```
Comment (yaqwsx) [inline RoFILib/snake_reconfig/MinMaxHeap.h:226]: inlines are unnecessary here and below.

File: RoFILib/snake_reconfig/Snake_algorithms.h [hunk 11]
```diff
@@ -0,0 +1,280 @@
...
+
+            if (newEval == 0) {
+                auto path = createPath(pred, pointerNext);
+                if (stat != nullptr) {
+                    stat->pathLength = path.size();
+                    stat->queueSize = maxQSize;
+                    stat->seenCfgs = pool.size();
+                }
+                return path;
+            }
```
Comment (yaqwsx) [inline RoFILib/snake_reconfig/Snake_algorithms.h:117]: The body is hard to follow. However, it is for a discussion on how to split it in pices.

**Comment by yaqwsx** [comment]
Thank you for all the work you put into it.

File: RoFILib/configuration/Configuration.cpp [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+    if (val > 180)
+        return 360 - val;
+
+    if (val <= -180)
+        return 360 + val;
+    
+    return val;
+}
+
+bool Module::setJoint(Joint joint, double val) {
```
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.cpp:49]: The boolean stand for "was the module/configuration changed?"

I think that it's mainly used in `Configuration::execute` to decide whether we should set `matricesVal` to `Value::Unknown`. I would leave it as it is now, but in the future it is needed to change the logic behind recomputing matrices.

File: RoFILib/configuration/Configuration.cpp [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+    modules.emplace(std::piecewise_construct,
+                    std::forward_as_tuple(id),  // args for key
+                    std::forward_as_tuple(alpha, beta, gamma, id));
+    edges.emplace(std::piecewise_construct,
+                    std::forward_as_tuple(id),  // args for key
+                    std::forward_as_tuple());
+    if ((modules.size() == 1) || (id < fixedId)) {
+        //TODO: je toto opravdu potřeba/proč to chceme
+        fixedId = id;
+    }
```
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.cpp:247]: I wanted to talk about this -- why we want to change the fixedId to the smallest one? I understand the `modules.size() == 1` part, but really dont see the `id < fixedId` one.

File: RoFILib/configuration/Configuration.cpp [hunk 14]
```diff
@@ -0,0 +1,496 @@
...
+
+bool Configuration::computeMatricesRec(ID id, ShoeId side) {
+    bool result = true;
+
+    // At first, fix other side of the module.
+    ShoeId side2 = side == A ? B : A;
+    auto& matrixCurr = matrices.at(id);
+    matrixCurr[side2] = computeOtherSideMatrix(id, side);
+
+    for (const std::optional<Edge>& edgeOpt : edges.at(id)) {
```
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.cpp:307]: In my opinion, yes. This is a private method and it's called with `fixedId` in `computeMatrices` and with `idNext` in `computeMatricesRec`, where it is called only on ids that are connected by an edge from current `id`

File: RoFILib/configuration/Configuration.cpp [hunk 15]
```diff
@@ -0,0 +1,496 @@
...
+    return mass;
+}
+
+bool Configuration::execute(const Action& action) {
+    bool ok = true;
+    for (const Action::Rotate rot : action.rotations())
+        ok &= execute(rot);
+
+    for (const Action::Reconnect rec : action.reconnections())
+        ok &= execute(rec);
```
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.cpp:404]: I am not sure, how it is used in Rofilib/distribute and in rrt, but in other parts of repo it would make sense.
So should I change it?

File: RoFILib/configuration/Configuration.cpp [hunk 16]
```diff
@@ -0,0 +1,496 @@
...
+            continue;
+
+        const Module &module = modules.at(id);
+        for (Joint joint : { Alpha, Beta, Gamma }) {
+            auto val = otherModule.getJoint(joint) - module.getJoint(joint);
+            if (joint == Gamma) {
+                val = clampGamma(val);
+            }
+            if (val != 0) {
+                rotations.emplace_back(id, joint, val);
```
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.cpp:425]: Could you explain, what do you mean exactly?

File: RoFILib/configuration/Configuration.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (yaqwsx) [reply RoFILib/configuration/Configuration.cpp:49]: Ok, please add a short comment explaining the return value now to not forget this.

File: RoFILib/configuration/Configuration.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (yaqwsx) [reply RoFILib/configuration/Configuration.cpp:247]: When you add the first module (`modules.size() == 1`), `fixedId` is undefined. Therefore you give it an initial value.

File: RoFILib/reconfig/rrt.cpp (hunk 9 above)
Comment (yaqwsx) [reply RoFILib/reconfig/rrt.cpp:21]: I changed my mind about anonymous namespace after reading this tweet: https://twitter.com/tvaneerd/status/1248684436796903426

File: RoFILib/configuration/Configuration.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (yaqwsx) [reply RoFILib/configuration/Configuration.cpp:307]: Ok then.

File: RoFILib/configuration/Configuration.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (yaqwsx) [reply RoFILib/configuration/Configuration.cpp:404]: I would change it - but consult it with tests, please.

File: RoFILib/configuration/Configuration.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.cpp:247]: Yes. But I don't understand the `id < fixedId` part, why is this the wanted behaviour?

[the remark this answers is not in the document]
Comment (yaqwsx) [reply RoFILib/configuration/Configuration.cpp:247]: Viki specified the module with the smallest id is the fixed one.

[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.cpp:247]: Oh, I've already broke this in my part of code. I use it as indicator "this is the root of the tree". So do we want to keep that? I feel like in future it would be better to let the user decide which module is the fixed in order to optimize recomputing `matrices`.

File: RoFILib/configuration/Configuration.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.cpp:404]: Seems that `distribute` doesn't have tests.

File: RoFILib/configuration/Configuration.h [hunk 17]
```diff
@@ -180,43 +92,12 @@ class Edge
     double onCoeff_ = 1;     //for visualizer
```
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.h:92]: Maybe @xbarnat will change it in his branch.

File: RoFILib/configuration/Configuration.h [hunk 18]
```diff
@@ -137,11 +63,12 @@ enum ShoeId { A = 0, B = 1 };
 enum ConnectorId { XPlus = 0, XMinus = 1, ZMinus = 2 };
 enum Orientation { North = 0, East = 1, South = 2, West = 3 };
 
-class Edge
-{
+class Edge {
```
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.h:66]: Honestly I am not used to working with structs, so I can only act as a rubber duck ¯\_(ツ)_/¯

File: RoFILib/configuration/Generators.h (hunk 1 above)
Comment (Trphajzlik) [reply RoFILib/configuration/Generators.h:9]: My bad, result of these functions being methods previously

File: RoFILib/configuration/Configuration.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (yaqwsx) [reply RoFILib/configuration/Configuration.cpp:247]: I agree with the future. Leave it as it is right now.

File: RoFILib/configuration/Configuration.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (yaqwsx) [reply RoFILib/configuration/Configuration.cpp:404]: Let's keep it as it this then. I don't want to break `distribute` as there is no one to fix it and bug finding in the future might be hard.

File: RoFILib/generate/main.cpp (hunk 2 above)
Comment (Trphajzlik) [reply RoFILib/generate/main.cpp:109]: No idea

File: RoFILib/configuration/Configuration.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (yaqwsx) [reply RoFILib/configuration/Configuration.h:66]: For now, leave it as it is.

File: RoFILib/generate/main.cpp (hunk 2 above)
Comment (yaqwsx) [reply RoFILib/generate/main.cpp:109]: Could you look at it? I don't want to keep commented conditions in the code.

File: RoFILib/configuration/Configuration.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (Trphajzlik) [reply RoFILib/configuration/Configuration.cpp:404]: It might be worth testing the impact of this change on efficiency of the algorithms based on astar, but I'll leave it as it this for now. Could you open issue about this?

File: RoFILib/snake_reconfig/Snake_algorithms.h (hunk 11 above)
Comment (Trphajzlik) [reply RoFILib/snake_reconfig/Snake_algorithms.h:117]: Could we leave this change for future work on my part of the code? (For new branch after merge)

File: RoFILib/configuration/Configuration.h [hunk 19]
```diff
@@ -36,8 +36,24 @@ class Module {
     ID getId() const { return id; }
     double getJoint(Joint a) const;
 
-    // return value specifies if the module was changed
+    /**
+     * \brief Rotates joint
+     * 
+     * \return `false` if adding \p val makes value of \p joint 
+     * overflow out of accepted range and value of joint isn't changed
```
Comment (yaqwsx) [inline RoFILib/configuration/Configuration.h:43]: Better to put it in a separate sentence in the middle of body. "Failure in this method leaves the joint unchanged"

File: RoFILib/configuration/Configuration.h [hunk 20]
```diff
@@ -236,16 +274,60 @@ class Configuration
...
 
     bool collisionFree() const;
 
     Vector massCenter() const;
 
+    /**
+     * \brief Tries to execute given \p action.
+     * 
+     * `False` is returned if either you try to add edge to already occupied
+     * dock, erase non-existent edge or if `rotate_joint` return `False`.
```
Comment (yaqwsx) [inline RoFILib/configuration/Configuration.h:324]: Do not refer to implementation details (a reference to \rotate_joint`) - better state "perform invalid rotation".

File: RoFILib/configuration/Configuration.h [hunk 21]
```diff
@@ -236,16 +274,60 @@ class Configuration
...
+     * 
+     * This function does NOT check two modules collide during rotations.
```
Comment (yaqwsx) [inline RoFILib/configuration/Configuration.h:326]: "check if"
